### PR TITLE
Add StatefulSet reconciler unit test for resize after scale-to-zero

### LINKED ISSUE #15300: Add StatefulSet reconciler unit test for resize after scale-to-zero
## Summary

Add unit-test coverage for the StatefulSet workload resize case that this change handles.

## Rationale

Pull request #15279 adds an integration test for scaling a Kueue-managed StatefulSet to zero and then scaling it up to a different replica count. Add a focused unit test to validate the StatefulSet reconciler behavior without requiring the integration test suite.

## Affected area

- `pkg/controller/jobs/statefulset/statefulset_reconciler_test.go`

## Required changes

Add a test case near the existing StatefulSet reconciler tests. The test must cover a StatefulSet that scales to zero and then scales up to a different replica count. It must verify that the Workload PodSet count is updated to the new replica count before the Workload is released from hold.

## Acceptance criteria

- The unit test covers scaling from zero to a different replica count.
- The unit test verifies the Workload PodSet count uses the current StatefulSet replica count.
- The unit test verifies the Workload can leave the on-hold state after the PodSet count is updated.

## Backlinks

- Originating pull request: https://github.com/kubernetes-sigs/kueue/pull/15279
- Originating review: https://github.com/kubernetes-sigs/kueue/pull/15279#pullrequestreview-5145555848
- Requested by: @tenzen-y

### PR DESCRIPTION
**Pull request by kavix**
#### What type of PR is this?
/kind cleanup
/area testing

#### What this PR does / why we need it?
Fixes #15300

Follow-up to #15279: adds unit-test coverage for the StatefulSet reconciler when scaling down to zero and then scaling up to a different replica count, ensuring the Workload PodSet count is updated before the hold is released.

#### Useful notes for your reviewer
Validates the reconciler behavior in `pkg/controller/jobs/statefulset/statefulset_reconciler_test.go` without requiring the integration test suite.

#### Release note
```release-note
NONE
```

## AI summary

- Adds StatefulSet reconciler coverage for scaling from zero to five replicas.
- Verifies the Workload PodSet count updates before release from hold.
- Tests both `UnadmittedWorkloadsObservability` settings and `QuotaReserved` conditions.

### Release note assessment

No contributor-supplied `release-note` block was found. Use:

```release-note
NONE
```

This change adds unit-test coverage only. Please verify this alternative and update the canonical `release-note` block in the PR description.

### CODE REVIEW

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go [hunk 1]
```diff
@@ -933,6 +935,128 @@ func TestReconciler_ClearOnHoldSetsReason(t *testing.T) {
 	}
 }
 
+func TestReconciler_ResizeAfterScaleToZero(t *testing.T) {
```
Comment (mimowo) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go]: Could we instead consolidate the new test cases into "TestReconciler", and instead introduce a generic param like "featureGates"?

Comment (kavix) [reply pkg/controller/jobs/statefulset/statefulset_reconciler_test.go]: Done! Consolidated the test cases directly into the `TestReconciler` table using the existing `featureGates` parameter.

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go [hunk 2]
```diff
@@ -68,17 +70,24 @@ func TestReconciler(t *testing.T) {
...
+		stsKey                      client.ObjectKey
+		statefulSet                 *appsv1.StatefulSet
+		pods                        []corev1.Pod
+		workloads                   []kueue.Workload
+		wantStatefulSet             *appsv1.StatefulSet
+		wantPods                    []corev1.Pod
+		wantWorkloads               []kueue.Workload
+		wantEvents                  []utiltesting.EventRecord
+		wantErr                     error
+		interceptorFuncs            interceptor.Funcs
```
Comment (tenzen-y) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go]: Remove interceptorFuncs from table and directly define it when fake client is initialized which means all test cases use the same client and interceptor.

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go [hunk 3]
```diff
@@ -251,6 +260,145 @@ func TestReconciler(t *testing.T) {
 					Obj(),
 			},
 		},
+		"should resize workload podSets when StatefulSet scales up from zero": {
+			featureGates: map[featuregate.Feature]bool{features.TopologyAwareScheduling: false},
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:264]:
```suggestion
```

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go [hunk 4]
```diff
@@ -251,6 +260,145 @@ func TestReconciler(t *testing.T) {
...
+						Status:  metav1.ConditionFalse,
+						Reason:  "NoReservation",
+						Message: "The workload has no reservation",
+					}).
+					Obj(),
+			},
+		},
+		"should resize workload podSets when StatefulSet scales up from zero with UnadmittedWorkloadsObservability disabled": {
+			featureGates: map[featuregate.Feature]bool{
+				features.TopologyAwareScheduling:          false,
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:336]:
```suggestion
```

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go [hunk 5]
```diff
@@ -251,6 +260,145 @@ func TestReconciler(t *testing.T) {
...
+			afterReconcile: func(t *testing.T) {
+				if !updatedBeforeHoldReleased.Load() {
+					t.Errorf("expected Workload PodSet count to be updated to new replica count before hold is released")
+				}
+			},
+			wantStatefulSet: statefulsettesting.MakeStatefulSet("sts", "ns").
+				UID("sts-uid").
+				Queue("lq").
+				Replicas(5).
+				DeepCopy(),
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:308]:
```suggestion
				Obj(),
```

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go [hunk 6]
```diff
@@ -251,6 +260,145 @@ func TestReconciler(t *testing.T) {
...
+			afterReconcile: func(t *testing.T) {
+				if !updatedBeforeHoldReleasedObservabilityDisabled.Load() {
+					t.Errorf("expected Workload PodSet count to be updated to new replica count before hold is released")
+				}
+			},
+			wantStatefulSet: statefulsettesting.MakeStatefulSet("sts", "ns").
+				UID("sts-uid").
+				Queue("lq").
+				Replicas(5).
+				DeepCopy(),
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:382]:
```suggestion
				Obj(),
```

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go [hunk 7]
```diff
@@ -68,17 +70,24 @@ func TestReconciler(t *testing.T) {
...
+		statefulSet                 *appsv1.StatefulSet
+		pods                        []corev1.Pod
+		workloads                   []kueue.Workload
+		wantStatefulSet             *appsv1.StatefulSet
+		wantPods                    []corev1.Pod
+		wantWorkloads               []kueue.Workload
+		wantEvents                  []utiltesting.EventRecord
+		wantErr                     error
+		interceptorFuncs            interceptor.Funcs
+		statefulSetForNextReconcile *appsv1.StatefulSet
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:89]: I'm not sure I fully understand the reason for this. Would it be possible to keep only the `statefulSet` and split the test into two cases: `scaling down to zero` and another one `scaling back up from zero`? I think that might keep the test simpler.

File: pkg/controller/jobs/statefulset/statefulset_reconciler_test.go [hunk 8]
```diff
@@ -251,6 +260,145 @@ func TestReconciler(t *testing.T) {
...
+			interceptorFuncs: interceptor.Funcs{
+				Update: func(ctx context.Context, c client.WithWatch, obj client.Object, opts ...client.UpdateOption) error {
+					if w, ok := obj.(*kueue.Workload); ok {
+						if len(w.Spec.PodSets) > 0 && w.Spec.PodSets[0].Count == 5 && workload.IsOnHold(w) {
+							updatedBeforeHoldReleased.Store(true)
+						}
+					}
+					return c.Update(ctx, obj, opts...)
+				},
+			},
```
Comment (mbobrovskyi) [inline pkg/controller/jobs/statefulset/statefulset_reconciler_test.go:298]: What exactly are we trying to verify here? Would it be sufficient to check that the workload conditions have been updated?
